### PR TITLE
[improve](agg) support distinct agg node

### PR DESCRIPTION
**Pull request by zhangstar333**
## Proposed changes
select c_name from customer union select c_name from customer
this sql used agg node to get distinct row of c_name,
so it's no need to wait for inserted all data to hash map,
could output the data which it's inserted into hash map successed.

Issue Number: close #xxx

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by zhangstar333** [comment]
run buildall

**Comment by zhangstar333** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45.02 seconds
 stream load tsv:          510 seconds loaded 74807831229 Bytes, about 139 MB/s
 stream load json:         21 seconds loaded 2358488459 Bytes, about 107 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.2 seconds inserted 10000000 Rows, about 342K ops/s
 storage size: 17161351195 Bytes

File: be/src/exec/exec_node.cpp [hunk 1]
```diff
@@ -371,7 +372,13 @@ Status ExecNode::create_node(RuntimeState* state, ObjectPool* pool, const TPlanN
         return Status::OK();
 
     case TPlanNodeType::AGGREGATION_NODE:
-        *node = pool->add(new vectorized::AggregationNode(pool, tnode, descs));
+        if (tnode.agg_node.__isset.use_streaming_preaggregation &&
+            tnode.agg_node.use_streaming_preaggregation &&
```
Comment (HappenLee) [inline be/src/exec/exec_node.cpp:376]: maybe do not need check stream preagg

File: be/src/vec/exec/distinct_vaggregation_node.cpp [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+    int rows = in_block->rows();
+    IColumn::Selector distinct_row;
+    distinct_row.reserve(rows);
+
+    RETURN_IF_CATCH_EXCEPTION(
+            _emplace_into_hash_table_to_distinct(distinct_row, key_columns, rows));
+
+    bool mem_reuse = _make_nullable_keys.empty() && out_block->mem_reuse();
+    if (mem_reuse) {
```
Comment (HappenLee) [inline be/src/vec/exec/distinct_vaggregation_node.cpp:87]: should use build_mutable_block to replace the code

File: be/src/vec/exec/distinct_vaggregation_node.cpp [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+class ObjectPool;
+} // namespace doris
+
+namespace doris::vectorized {
+
+DistinctAggregationNode::DistinctAggregationNode(ObjectPool* pool, const TPlanNode& tnode,
+                                                 const DescriptorTbl& descs)
+        : AggregationNode(pool, tnode, descs) {}
+
+Status DistinctAggregationNode::get_next(RuntimeState* state, Block* block, bool* eos) {
```
Comment (HappenLee) [inline be/src/vec/exec/distinct_vaggregation_node.cpp:34]: no need support in not pipeline query engine

File: be/src/pipeline/exec/distinct_streaming_aggregation_sink_operator.cpp [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+                                              SourceState source_state) {
+    if (in_block && in_block->rows() > 0) {
+        if (_output_block == nullptr) {
+            _output_block = _data_queue->get_free_block();
+        }
+        RETURN_IF_ERROR(_node->do_pre_agg(in_block, _output_block.get()));
+
+        // get enough data or reached limit rows, need push block to queue
+        if (_output_block->rows() >= state->batch_size() ||
+            (_node->limit() != -1 && _output_block->rows() >= _node->limit())) {
```
Comment (HappenLee) [inline be/src/pipeline/exec/distinct_streaming_aggregation_sink_operator.cpp:56]: _output_block->rows() >= _node->limit() ？

File: be/src/pipeline/exec/distinct_streaming_aggregation_sink_operator.cpp [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+            reinterpret_cast<vectorized::DistinctAggregationNode*>(_node)
+                    ->update_output_distinct_rows(_output_block->rows());
+            _data_queue->push_block(std::move(_output_block));
+        }
+    }
+
+    // reach limit or source finish
+    if ((UNLIKELY(source_state == SourceState::FINISHED)) ||
```
Comment (HappenLee) [inline be/src/pipeline/exec/distinct_streaming_aggregation_sink_operator.cpp:64]: here should cut the output block rows to match the limit

File: be/src/vec/exec/distinct_vaggregation_node.cpp (hunk 2 above)
Comment (zhangstar333) [reply be/src/vec/exec/distinct_vaggregation_node.cpp:63]: have some problem to use the function,  maybe keep this as agg node  [                      if (!mem_reuse) {](https://github.com/apache/doris/blob/527547b4ed14cdf524576755961952119f12bc06/be/src/vec/exec/vaggregation_node.cpp#L1121)

**Comment by zhangstar333** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 44.98 seconds
 stream load tsv:          503 seconds loaded 74807831229 Bytes, about 141 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          29.3 seconds inserted 10000000 Rows, about 341K ops/s
 storage size: 17169455991 Bytes

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45.92 seconds
 stream load tsv:          506 seconds loaded 74807831229 Bytes, about 140 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          66 seconds loaded 1101869774 Bytes, about 15 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.0 seconds inserted 10000000 Rows, about 344K ops/s
 storage size: 17168768454 Bytes

**Comment by zhangstar333** [comment]
run buildall

**Comment by zhangstar333** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 47.42 seconds
 stream load tsv:          503 seconds loaded 74807831229 Bytes, about 141 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.2 seconds inserted 10000000 Rows, about 342K ops/s
 storage size: 17168744223 Bytes

**Comment by zhangstar333** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 44.69 seconds
 stream load tsv:          503 seconds loaded 74807831229 Bytes, about 141 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          66 seconds loaded 1101869774 Bytes, about 15 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          29.3 seconds inserted 10000000 Rows, about 341K ops/s
 storage size: 17169223037 Bytes

**Comment by zhangstar333** [comment]
run feut

**Review by HappenLee — APPROVED** [review]
LGTM

**Comment by zhangstar333** [comment]
run feut

**Comment by zhangstar333** [comment]
run feut

**Comment by zhangstar333** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 44.55 seconds
 stream load tsv:          503 seconds loaded 74807831229 Bytes, about 141 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          29.3 seconds inserted 10000000 Rows, about 341K ops/s
 storage size: 17163093419 Bytes

**Comment by zhangstar333** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 45.51 seconds
 stream load tsv:          508 seconds loaded 74807831229 Bytes, about 140 MB/s
 stream load json:         19 seconds loaded 2358488459 Bytes, about 118 MB/s
 stream load orc:          65 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          30 seconds loaded 861443392 Bytes, about 27 MB/s
 insert into select:          29.5 seconds inserted 10000000 Rows, about 338K ops/s
 storage size: 17163492937 Bytes

**Review by BiteTheDDDDt — APPROVED** [review]

**Pull request merged**